From ef3e9e6dbc38c76da4202beefd93e4f0a660a14b Mon Sep 17 00:00:00 2001
From: Ewan Milne <emilne@redhat.com>
Date: Wed, 20 Aug 2014 18:59:16 -0400
Subject: [scsi] scsi_debug: invalidate protection info for unmapped region

Message-id: <1408561176-18133-38-git-send-email-emilne@redhat.com>
Patchwork-id: 88460
O-Subject: [RHEL7.1 PATCH e-stor 37/57] [SCSI] scsi_debug: invalidate protection info for unmapped region
Bugzilla: 1132100
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: "Ewan D. Milne" <emilne@redhat.com>

upstream commit e9926b4376544d5a2dc4d310d4d0006c634b1a93

Tested by Ewan D. Milne and Barry Donahue with a combination
of manual and automated testing.  No issues seen.

upstream commit e9926b4376544d5a2dc4d310d4d0006c634b1a93
Author: Akinobu Mita <akinobu.mita@gmail.com>
Date:   Sat Jun 29 17:59:17 2013 +0900

    [SCSI] scsi_debug: invalidate protection info for unmapped region

    When UNMAP command is issued with the data integrity support enabled,
    the protection info for the unmapped region is remain unchanged.
    So READ command for the region later on causes data integrity failure.

    This fixes it by invalidating protection info for the unmapped region
    by filling with 0xff pattern.

    Signed-off-by: Akinobu Mita <akinobu.mita@gmail.com>
    Acked-by: Douglas Gilbert <dgilbert@interlog.com>
    Acked-by: "Martin K. Petersen" <martin.petersen@oracle.com>
    Signed-off-by: James Bottomley <JBottomley@Parallels.com>

Signed-off-by: Ewan D. Milne <emilne@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/scsi_debug.c b/drivers/scsi/scsi_debug.c
index 9eb6303..df4e1da 100644
--- a/drivers/scsi/scsi_debug.c
+++ b/drivers/scsi/scsi_debug.c
@@ -2064,6 +2064,11 @@ static void unmap_region(sector_t lba, unsigned int len)
            scsi_debug_sector_size *
            scsi_debug_unmap_granularity);
    }
+   if (dif_storep) {
+    memset(dif_storep + lba, 0xff,
+           sizeof(*dif_storep) *
+           scsi_debug_unmap_granularity);
+   }
   }
   lba = map_index_to_lba(index + 1);
  }
-- 
1.7.1